This could be called... TextPlus2!

ToriumSlurs opened the pull request:
This PR modularizes the entire codebase and separates functionality into separate modules for improved maintainability. This refactor was necessary, as TextPlus was approaching 2k lines of code. It also introduces performance enhancements to improve runtime efficiency.  

# PLEASE NOTE:
Most basic features have been tested to the extent of my availability and are working fine (custom fonts haven't been tested and some functions that I personally would rarely use)

ToriumSlurs committed:
Modularize entire code, Separate into separate modules, Performance enchancements

AlexanderLindholt commented:
Hi there!

First, I'd like to say that I really appreciate your effort in helping improve the library. That said, I do have some feedback regarding your changes.

In programming — as in most development environments — many choices are a matter of preference, and not everyone will always agree. While this is an open-source project (and passionately so), I'd like to avoid major refactorings that greatly alter the coding style, so that the original style remains.

I'm always happy to accept contributions, and I often implement features and improvements myself after checking out contributor solutions. However, in this case, your update seems to be mostly a huge modularization and stylistic shift, with no functional changes.

Regarding optimization, it seems that your changes might actually hurt performance compared to the original version, despite your claims. I'd be very interested if you could prove me wrong.

I hope this clears things up. Make sure to let me know if I missed anything as well.

ToriumSlurs committed:
Update FontManager.luau

Adding minor type annotations

ToriumSlurs committed:
Update InstanceCache.luau

Adding minor type annotations + optimizing cache

ToriumSlurs committed:
Update init.luau

Adding minor type annotations

ToriumSlurs commented:
> Regarding optimization, it seems that your changes might actually hurt performance compared to the original version, despite your claims. I'd be very interested if you could prove me wrong.

You were initially right, now that I have optimized the InstanceCache further the performance is better by just a little bit.
<img width="1332" height="466" alt="image_2025-09-28_001239303" src="https://github.com/user-attachments/assets/64e7742a-aa5c-4ea4-b2f0-ca79eae2fa37" />

ToriumSlurs committed:
Update init.luau

big oops

ViliamOleos commented:
# Hey people, I noticed this about 20 minutes after getting posted, but after trying to reach @AlexanderLindholt for a short amount of time, reviewing the whole code, finding things, thinking and then writing this whole thing, I kinda don't want to adjust it all to the updated conversation (huge sorry), so I'm just posting it like that and I'm going to respond to the updates in a separate message.

Okay, this is kind of disastrous. I'm going to try to get a hold of Alexander, but as for right now, I'm going to try to review how this could impact the codebase.

## Runtime Code Compilation

Okay, so You know how in every script of this, You've got `--!optimize 2` and `--!native` at the start? 
Well, see, even off the top of my head, I know that's not the greatest thing, 'cause I've read the Roblox Docs, but here's a quote: _"It may be tempting to place the `--!native` comment in every script just in case some of them will execute faster, but native code generation has some drawbacks:"_

And, yeah, maybe we could ignore some of those, but here's the one that stings the most: _"There's a limit on the total allowed amount of natively compiled code in an experience."_
I admit, I do not know the extent and the metric by how this is done, but I've done some experiments before, and while I do not remember the results exactly, I've learned that the limit for these things is relatively _VERY_, _VERY_ small. 

I understand that this is not concrete evidence, and no one really knows for sure what happens in Roblox's backend, but it could be worrying if that's true.
Imagine if someone has a bunch of modules that use native code compilation and then Text+ _"2"_ comes in and produces this chonking honk of compiled scripts, that would cause a lot of frustration.

## Lack of Documentation

For seemingly such a big pull request, there's nothing written about it! All we got is _"this modularizes the entire codebase nothing to talk about, this has performance improvements but im not telling you what they are.. OH and btw this is necessary"_. No proof or theory of performance, no nothing. That's what this implied, even if You honestly don't mean that.

For all I know, this could be vibe-coded. Looking at the code, it doesn't seem to be, but still.

### "Project Coup"

To absolutely safely merge this, we would have to review every single line of code (and extremely likely provide edits!!) basically from scratch, and since there is no aforementioned documentation (including the lack of comments), there's no help describing it, all for the end result of mostly the same functionality, but _"cooler"_.

And merging this without thinking or editing leads to what I like to call _"Project Coup"_, which is a thing that destroyed many open-source projects. Here's how a _"Project Coup"_ usually goes like this:

<details><summary>Open this up if You'd like a read about that.</summary>
<p>

`----------------------------------------------------------------------`

- Some new person comes into the project, rewrites a bunch of it and it's supposedly "better" (sometimes it is, sometimes it's not). The worst cases are done in their own style.
- The maintainers (more likely to occur with private projects, with less or one maintainers) happily accept the changes, liking the benefits while not caring enough to completely check, edit or understand the code.
- (In some cases the new person continues to rewrite a bunch of the codebase/add new code but not in every case)
- Eventually the original maintainers struggle to understand and maintain the code (and/or perhaps there are problems in the rewritten code)
- And the new person basically becomes the primary maintainer, basically _"couping"_ the project for themselves. After a while, it's likely that most of the original maintainers silently give up on the project.
- Maybe the new person never did any other changes, left the project or went in a hibernation, but the original maintainers come back to the project and they either have to reverse all the commits done after the new person or rewrite a bunch of the code.

So yeah, not a very desirable series of events. This happened to projects, and in some cases the original maintainers may never come back to clean up the mess and continue the project after they've lost motivation.

I'm not saying this is the inevitable thing that happens if we just merge this commit, but the main morals are that we should be more accurate accepting code from strangers, that we should follow coding guidelines and that we need to properly document our code.

`----------------------------------------------------------------------`

</p>
</details> 

## A thing about this specific codebase.

As far as I know, this codebase doesn't really appreciate pull requests. If something's in a pull request that could be useful, it's getting rewritten separately and closed, rather than edited officialy or something like that. That's why I write issues with code suggestions here now, not making changes and presenting them for maximum help.

This is not gonna run 100%, especially such a big one.

## Minor nitpicks

These are some mroe minor nitpicks I didn't consider worthy enough to be a full section. These aren't as important, but still wanted to mention them and provide some criticism.

- The main file now is not called _"TextPlus"_, that's been deleted. It's now replaced by _"init"_. This may not seem to be too big of a deal, but I'd say that's _"ripping the face off"_, because now, to use the module, You've gotta reference some undescriptive _"init"_ script, and in Explorer it would NOT stand out in the probable sea of other module scripts the dev is using along Text+. Yes, You could put all the code in a folder, but I'm not sure if that's the intention and that's still going to make including Text+ in code a bit weirder and less intuitive, even if by a small amount.
- I see right now that the branch You're merging this from is the `main` from Your fork. To be honest, nothing too wrong with that, I merged from `main` too before I learned this, but the standard way to do it is to make a separate branch for the pull request in Your own fork, and then merge off of that. It's a lot more comfortable in many ways.

## Closing thoughts

The idea isn't particularly bad, and I see You are well intentioned, but a lot of things are not helping the case, I hope You understand, mate.

And with the thing of code compilation, it's hard to even imagine a more modular codebase. It absolutely could be possible, though, I have an idea, but it's pretty esoteric, I'm still going to pitch it for fun
- What if you could keep the source code in GitHub all modular and then have it "compiled" with specific tooling to a single script that then works in Roblox Studio? Just a thought.

Anyways, it has been a BEEG message, have an amazing day I guess.

ToriumSlurs commented:
> For seemingly such a big pull request, there's nothing written about it! All we got is "this modularizes the entire codebase nothing to talk about, this has performance improvements but im not telling you what they are.. OH and btw this is necessary". No proof or theory of performance, no nothing. That's what this implied, even if You honestly don't mean that.

There really has nothing to be written about it, the api stays the same, just modularized and improved in some areas such as introduction of the InstanceCache. Modularizing it makes it way easier to maintain for the creator, and for us, the users way easier to interact with, read through, learn and contribute in the future, because before it was all in a single script and to really get a good look at a particular part of the script you had to look through the 1800+ lines.

>The main file now is not called "TextPlus", that's been deleted. It's now replaced by "init". This may not seem to be too big of a deal, but I'd say that's "ripping the face off", because now, to use the module, You've gotta reference some undescriptive "init" script, and in Explorer it would NOT stand out in the probable sea of other module scripts the dev is using along Text+. Yes, You could put all the code in a folder, but I'm not sure if that's the intention and that's still going to make including Text+ in code a bit weirder and less intuitive, even if by a small amount.

The TextPlus module has been renamed to init because as you may know you cannot place other scripts under a script in github! In roblox studio you can, and in roblox studio the init would be named TextPlus and the other modules would be under it.

ToriumSlurs commented:
>I see, it's quite a nice improvement, I'm usually a very performance-centered person, but to be absolutely honest, if this pull request really gives us the migraines that I assume it gives us, it's not worth over something that could be very well much more easily implemented in the One Big Page variant of Text+

There is no migraines. Just drag and drop. Works the same, just is cleaner and slightly more performant.

AlexanderLindholt commented:
> slightly more performant.

You have yet to document your benchmark and prove the results.
It doesn't logically make sense at first glance due to the obvious performance hurt from modularization.

ToriumSlurs commented:
> You have yet to document your benchmark and prove the results.

Yet another benchmark:
<img width="1571" height="898" alt="7dc929612a32b10162d14731bc283803" src="https://github.com/user-attachments/assets/1a902a17-2575-444a-afb5-a3f5d85b9f4b" />

The benchmark code:
```lua
--[[
This file is for use by Benchmarker (https://boatbomber.itch.io/benchmarker)

|WARNING| THIS RUNS IN YOUR REAL ENVIRONMENT. |WARNING|
--]]

local StarterGui = game.StarterGui
local CoreGui = StarterGui.Core
local AlertFrame = CoreGui.Alerts
local TextFrame = AlertFrame.TextFrame

return {
	ParameterGenerator = function()
		return
	end,

	Functions = {
		["TextPlusModularized"] = function(Profiler)
			local TextPlusModularized = require(game.ReplicatedStorage.TextPlusModularized)
			local TestText = TextPlusModularized.Create(TextFrame, `Test Test Test Test Test Test Test Test Test Test Test Test Test Test Test Test Test Test`, 
				{
					Font = Font.new(
						"rbxasset://fonts/families/Guru.json",
						Enum.FontWeight.Regular,
						Enum.FontStyle.Normal
					),
					Size = 50,
					Color = Color3.fromRGB(192, 24, 0),
					StrokeSize = 2,
					WordSorting = true,
					StrokeColor = Color3.fromRGB(70, 3, 0),
					StrokeTransparency = 1,
					Dynamic = true,
					XAlignment = "Center",
					YAlignment = "Center",
					Transparency = 1,
				}
			)
			
			local bounds = TextPlusModularized.GetBounds(TextFrame)
			local text = TextPlusModularized.GetText(TextFrame)
			local options = TextPlusModularized.GetOptions(TextFrame)

			for _, character in TextPlusModularized.GetCharacters(TextFrame) do
				character.TextColor3 = Color3.fromRGB(255, 255, 255)
				character.TextTransparency = 1
			end
		end,

		["TextPlus"] = function(Profiler) 
			local TextPlus = require(game.ReplicatedStorage.TextPlus)
			local TestText = TextPlus.Create(TextFrame, `Test Test Test Test Test Test Test Test Test Test Test Test Test Test Test Test Test Test`, 
				{
					Font = Font.new(
						"rbxasset://fonts/families/Guru.json",
						Enum.FontWeight.Regular,
						Enum.FontStyle.Normal
					),
					Size = 50,
					Color = Color3.fromRGB(192, 24, 0),
					StrokeSize = 2,
					WordSorting = true,
					StrokeColor = Color3.fromRGB(70, 3, 0),
					StrokeTransparency = 1,
					Dynamic = true,
					XAlignment = "Center",
					YAlignment = "Center",
					Transparency = 1,
				}
			)
			
			local bounds = TextPlus.GetBounds(TextFrame)
			local text = TextPlus.GetText(TextFrame)
			local options = TextPlus.GetOptions(TextFrame)
			
			for _, character in TextPlus.GetCharacters(TextFrame) do
				character.TextColor3 = Color3.fromRGB(255, 255, 255)
				character.TextTransparency = 1
			end
		end,
	},
}
```

>As you noted yourself, not all features have been tested nor are guaranteed to work.

Everything seems to be working correctly in the benchmark. I am using all of the functions in the api for it.

ViliamOleos commented:
@ToriumSlurs 
1. I do believe there is a bunch that could be written about it, but trying not to beat a dead horse around here, changing the code at all without close inspection and editing really does throw off the creator a bunch if I speak from my experience. But for the users, I completely agree. Not much to say about that.
2. And yeah, that was a nitpick, I wasn't really sure if that's exactly what You were intending with that code.

@AlexanderLindholt 
Okay, yeah, I always assumed that was intentional. And I do agree that over-modularization (and, in my opinion, over-abstraction) should be avoided. I like my code more packed too.

I appreciate it, and didn't really wanted to offend, I just chose the word _"spaghetti"_, it's not _horribly_ bad, but I've always had _some_ trouble with it.

And I did inspect some of Your other projects, but not too in-depth, and I especially never looked at Tween+, but I understand Your reasoning. 

### =====================================================

No migraines? I understand how that's a position You could take, but just consider for a moment what I described in the giant message. Could be potentially problematic.

ViliamOleos commented:
@ToriumSlurs I see. I'm not experienced in that kind of Roblox benchmarking, so I can't say too much or make any assumptions, though.

AlexanderLindholt commented:
@ToriumSlurs Can you confirm that every single feature that you're using in the benchmark actually works properly? Dynamic mode, alignments, sorting, and strokes.

ToriumSlurs commented:
@AlexanderLindholt Yes I can confirm, I am using the modularized textplus module in a game of mine and everything is working correctly.

ViliamOleos commented:
@ToriumSlurs I see, that is pretty convincing if that's true. Still, it's not possible to confirm every edge case works correctly, so there's still some risk without standardised testing.

AlexanderLindholt commented:
@ToriumSlurs You have to share what might be the big improvements here, because it seems almost impossible if you consider all of your sacrifices.

ToriumSlurs commented:
> You have to share what might be the big improvements here

I split the large TextPlus module into several distinct modules. Once that was complete and free of errors, I refactored each one of them, simplifying, optimizing functions, introducing weak tables, and identifying code that was heavy on the CPU. I didn’t keep a detailed changelog, but I’ve summarized all the key improvements.

> Have you tried specifically the options used in the benchmark — see if everything works perfectly?

Like I said previously, everything works fine. Creation, Customization etc. NO ISSUES! :)

AlexanderLindholt commented:
> Yeah, it's definitely possible to optimize enough that the modularized version is faster

The thing is, modularization slows things down. If his version and his benchmarks really work as they should, I'm interested in what he did besides modularization.

> try to make small commits sometimes, making them descriptive enough, and then just list those commits of as a changelog.

I personally just write down every change or category of change I make when I'm done implementing it.

ToriumSlurs commented:
Doesn’t look as if much was changed…
<img width="590" height="1278" alt="IMG_7193" src="https://github.com/user-attachments/assets/b5adf9f7-0eb4-4289-a01d-9a2438deef51" />

I still hope my effort to modularize TextPlus won’t go to waste. Please do not judge a book by it’s cover. You keep saying that modularization slows it down and yet, it is faster than the normal module.

ToriumSlurs commented:
@ViliamOleos 
The examples you have provided do not actually address the core issue. While your examples make the modularised version appear longer, my approach in practice reduces both code length and complexity, as shown here:

<img width="149" height="22" alt="image_2025-09-28_132720913" src="https://github.com/user-attachments/assets/b0459ac3-8d05-41b6-be3a-a338d9f9b3c9" />

This clearly shows that my modularisation improves efficiency without adding unnecessary overhead or complexity.

Regarding compilation limits, my method adds no extra scope declarations or redundant runtime costs. These modules are intentionally lightweight, containing only things that are essential, which means the risk of hitting "too much code" errors is minimal.

Modularisation is about more than just performance, it allows isolated testing, simplifies maintenance, and reduces the likelihood of introducing bugs when changes are made. This aligns perfectly with the long-term goals of stability and scalability for Text+.

The idea that modularisation automatically increases complexity does not hold true here. In fact, my approach delivers clear benefits for both performance and maintainability.

I have thoroughly evaluated these changes and believe they add significant value. The real proof, however, comes from implementation. I encourage you to test this approach yourselves. If uncertainty remains, the most transparent solution is to **let the Text+ user base decide whether they prefer a single, tangled script or a clean, modular structure that supports efficiency, adaptability, and sustainability**.

AlexanderLindholt commented:
@ToriumSlurs I might also take a look at how I could modularize Text+ myself the best. I understand the valid concern that Text+ is getting huge. Thanks for your time so far.

ToriumSlurs committed:
quick fix (init.luau)

ViliamOleos commented:
Alright, good to hear. I saw that the new message still had a bunch of misunderstandings and thought I'll have to change up my strategy to figure out what was going on, but I guess it's best to just not agree and force a decision on.

Oh, and btw, don't worry too much about the changelog at this point, it's a dead horse by now. People make mistakes. And I wanted to mention that it doesn't have to be too detailed to death. Just enough to figure out what's going on, and if someone is confused, they can just ask!

Yes, manual modularisation is probably going to be the best modularisation, though my opinion is purely against any modularisation and that there are ways to drastically improve the single script readability without modularising it all))
That's just my opinion though.

AlexanderLindholt commented:
> I think that plugin is paid too? I'm not sure, but it would suck.

Yes, the plugin is paid. I bought it, found out it had a few issues and missing features, contacted the developer, informed him of the difficulties with the product, and despite promising to update it, he still hasn't pushed a single change after 6 whole months — ridiculous!

I’d recommend thinking twice before purchasing it. You're paying for a half-working plugin that isn't being maintained.

Not to mention that the code is super unreadable. :(

AlexanderLindholt commented:
> What are we doing then?

The only thing I can take from this pull request is that I should modularize the codebase to some degree.

It won't be to the same extent as was suggested though. It'll just be small readability improvements alongside the rest of what's going to be in version 1.29.
